Description: Include config in maintenance:install
 These changes ensure config is read in maintenance::install, enabling safe
 initialization of the server.
Author: Martin Habovstiak <martin.habovstiak@gmail.com>
Last-Update: 2023-10-08
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
Index: nextcloud-server-27.1.2/core/Command/Maintenance/Install.php
===================================================================
--- nextcloud-server-27.1.2.orig/core/Command/Maintenance/Install.php
+++ nextcloud-server-27.1.2/core/Command/Maintenance/Install.php
@@ -30,6 +30,9 @@
  */
 namespace OC\Core\Command\Maintenance;
 
+require_once __DIR__ . '/../../../lib/base.php';
+
+use OC;
 use bantu\IniGetWrapper\IniGetWrapper;
 use InvalidArgumentException;
 use OC\Installer;
@@ -113,6 +116,27 @@ class Install extends Command {
 		return 0;
 	}

+	protected function getOption(InputInterface $input, string $name) {
+		$option_mapping = [
+			'database' => 'dbtype',
+			'database-user' => 'dbuser',
+			'database-pass' => 'dbpassword',
+			'database-name' => 'dbname',
+			'database-host' => 'dbhost',
+			'admin-user' => 'adminlogin',
+			'admin-pass' => 'adminpass',
+			'admin-email' => 'adminemail',
+			'data-dir' => 'datadirectory',
+		];
+
+		// Ideally CLI args would override conf args not the other way around but Symfony doesn't
+		// expose an API to check if argument was given. >:(
+		if (in_array($option_mapping[$name], OC::$config->getKeys())) {
+			return OC::$config->getValue($option_mapping[$name]);
+		} else {
+			return $input->getOption($name);
+		}
+	}
 	/**
 	 * @param InputInterface $input
 	 * @param OutputInterface $output
@@ -120,33 +144,33 @@ class Install extends Command {
 	 * @return array
 	 */
 	protected function validateInput(InputInterface $input, OutputInterface $output, $supportedDatabases) {
-		$db = strtolower($input->getOption('database'));
+		$db = strtolower($this->getOption($input, 'database'));
 
 		if (!in_array($db, $supportedDatabases)) {
 			throw new InvalidArgumentException("Database <$db> is not supported. " . implode(", ", $supportedDatabases) . " are supported.");
 		}
 
-		$dbUser = $input->getOption('database-user');
-		$dbPass = $input->getOption('database-pass');
-		$dbName = $input->getOption('database-name');
-		$dbPort = $input->getOption('database-port');
+		$dbUser = $this->getOption($input, 'database-user');
+		$dbPass = $this->getOption($input, 'database-pass');
+		$dbName = $this->getOption($input, 'database-name');
+		$dbPort = $this->getOption($input, 'database-port');
 		if ($db === 'oci') {
 			// an empty hostname needs to be read from the raw parameters
 			$dbHost = $input->getParameterOption('--database-host', '');
 		} else {
-			$dbHost = $input->getOption('database-host');
+			$dbHost = $this->getOption($input, 'database-host');
 		}
 		if ($dbPort) {
 			// Append the port to the host so it is the same as in the config (there is no dbport config)
 			$dbHost .= ':' . $dbPort;
 		}
 		if ($input->hasParameterOption('--database-pass')) {
-			$dbPass = (string) $input->getOption('database-pass');
+			$dbPass = (string) $this->getOption($input, 'database-pass');
 		}
-		$adminLogin = $input->getOption('admin-user');
-		$adminPassword = $input->getOption('admin-pass');
-		$adminEmail = $input->getOption('admin-email');
-		$dataDir = $input->getOption('data-dir');
+		$adminLogin = $this->getOption($input, 'admin-user');
+		$adminPassword = $this->getOption($input, 'admin-pass');
+		$adminEmail = $this->getOption($input, 'admin-email');
+		$dataDir = $this->getOption($input, 'data-dir');
 
 		if ($db !== 'sqlite') {
 			if (is_null($dbUser)) {
Index: nextcloud-server-27.1.2/lib/base.php
===================================================================
--- nextcloud-server-27.1.2.orig/lib/base.php
+++ nextcloud-server-27.1.2/lib/base.php
@@ -130,7 +130,7 @@ class OC {
 
 	public static \OC\Server $server;
 
-	private static \OC\Config $config;
+	public static \OC\Config $config;
 
 	/**
 	 * @throws \RuntimeException when the 3rdparty directory is missing or
